docs/lang/articles/basic/syntax.md

```diff
@@ -4,280 +4,484 @@ sidebar_position: 1
+
     ...
 
+
+
 @ti.kernel
+
 def error_too_many_arguments(u: ti.i32, v: ti.i64, w: ti.types.matrix(7, 9, ti.i64)):  # Error: has 65 elements in total
+
     ...
 ```
 
-:::
+<details>
+
+<summary><font color="#006284">Advanced arguments</font></summary>
+
+*You can skip this part if you are just beginning.*
+
+A kernel can also take the following two types of advanced arguments:
+
+- Template arguments: Use `ti.template()` as their type hints. See [Template metaprogramming](../advanced/meta.md#template-metaprogramming). 
+- External array arguments: Use `ti.ext_arr()` as their type hints. See [Interacting with external arrays](./external.md).
+
+</details>
 
 ### Return value
 
-It is optional for a kernel to have a return value. If specified, it must be a type hinted **scalar/vector/matrix** value:
+A kernel can have *at most* one return value, which can be a scalar, `ti.Matrix`, and `ti.Vector`. Follow these rules when defining the return value of a kernel: 
 
-```python {2}
+- Type hint the return value of a kernel. 
+- Ensure that you have *at most* one return value in a kernel. 
+- Ensure that you have *at most* one return statement in a kernel. 
+- Ensure that the number of elements in the return value does not exceed 30. 
+
+#### Type hint the return value of a kernel
+
+```Python
 @ti.kernel
-def my_kernel() -> ti.f32:
-    return 128.32
 
-print(my_kernel())  # 128.32
+def test(x: ti.f32) -> ti.f32: # The return value is type hinted
+
+    return 1.0
 ```
 
-In addition, the return value will be automatically cast into the hinted type:
+In addition, the return value is automatically cast into the hinted type:
 
-```python {2-3,5}
+```Python
 @ti.kernel
+
 def my_kernel() -> ti.i32:  # int32
+
     return 128.32
 
-print(my_kernel())  # 128, cast into ti.i32
+print(my_kernel())  # 128, the return value is cast into ti.i32
 ```
 
-:::note
+#### At most one return value in a kernel
+
+```Python
+@ti.kernel
+
+def error_multiple_return() -> (ti.i32, ti.f32):
+
+    x = 1
+    y = 0.5
+    return x, y  # Compilation error: more than one return value
+```
 
-For now, a kernel can only have one return value, and the number of elements in the return value must not exceed 30.
+#### At most one return statement in a kernel
 
-```python {2,6,10}
+```Python
 @ti.kernel
-def valid_scalar_return() -> ti.f32:
-    return 128.0  # Return 128.0
 
+def test_sign(x):
+
+    if x >= 0:
+        return 1.0
+    else:
+        return -1.0
+    # Error: multiple return statements
+```
+
+As a workaround, you can save the result in a local variable and return it at the end:
+
+```Python
 @ti.kernel
-def valid_matrix_return() -> ti.types.matrix(2, 2, ti.i32):
-    return ti.Matrix([[1, 0], [0, 1]])
+
+def test_sign(x):
+
+    sign = 1.0
+
+    if x < 0:
+
+        sign = -1.0
+
+    return sign
+
+    # One return statement works fine
+```
+
+#### At most 30 elements in the return value
```

Comment:
This 30 seems arbitrary (I can understand it from a dev's perspective, tho). Is Taichi able to throw out a compiler-error hinting the users about this limitation? @lin-hitonami

Reply:
The error is already thrown.